server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -538,47 +616,209 @@ public void trimUnreferencedReaders() throws IOException {
         // store.
         if (startedPrimarySupplier.getAsBoolean() == false || pauseSync.get()) {
             return;
         }
 
+        // This is to fail fast and avoid listing md files un-necessarily.
+        if (RemoteStoreUtils.isPinnedTimestampStateStale()) {
+            logger.warn("Skipping remote segment store garbage collection as last fetch of pinned timestamp is stale");
+            return;
+        }
+
         // Since remote generation deletion is async, this ensures that only one generation deletion happens at a time.
         // Remote generations involves 2 async operations - 1) Delete translog generation files 2) Delete metadata files
         // We try to acquire 2 permits and if we can not, we return from here itself.
         if (remoteGenerationDeletionPermits.tryAcquire(REMOTE_DELETION_PERMITS) == false) {
             return;
         }
 
-        // cleans up remote translog files not referenced in latest uploaded metadata.
-        // This enables us to restore translog from the metadata in case of failover or relocation.
-        Set<Long> generationsToDelete = new HashSet<>();
-        for (long generation = minRemoteGenReferenced - 1 - indexSettings().getRemoteTranslogExtraKeep(); generation >= 0; generation--) {
-            if (fileTransferTracker.uploaded(Translog.getFilename(generation)) == false) {
-                break;
+        ActionListener<List<BlobMetadata>> listMetadataFilesListener = new ActionListener<>() {
+            @Override
+            public void onResponse(List<BlobMetadata> blobMetadata) {
+                List<String> metadataFiles = blobMetadata.stream().map(BlobMetadata::name).collect(Collectors.toList());
+
+                try {
+                    if (metadataFiles.size() <= 1) {
+                        logger.debug("No stale translog metadata files found");
+                        remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                        return;
+                    }
+
+                    // Check last fetch status of pinned timestamps. If stale, return.
+                    if (RemoteStoreUtils.isPinnedTimestampStateStale()) {
+                        logger.warn("Skipping remote segment store garbage collection as last fetch of pinned timestamp is stale");
+                        remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                        return;
+                    }
+
+                    List<String> metadataFilesToBeDeleted = getMetadataFilesToBeDeleted(metadataFiles);
+
+                    // If no metadata file is filtered out, make sure to keep latest metadata file
+                    if (metadataFiles.size() == metadataFilesToBeDeleted.size() && metadataFiles.equals(metadataFilesToBeDeleted)) {
+                        metadataFilesToBeDeleted = metadataFiles.subList(1, metadataFiles.size());
```

Comment:
do we need to ensure latest metadata file is kept along even if some files are filtered out in `getMetadataFilesToBeDeleted` ?

Reply:
Good point! It is possible that we filter out older files, this if condition fails and we end up deleting latest metadata file.

If we replace this if block with:
```
metadataFilesToBeDeleted.remove(metadataFiles.get(0));
```
would it work?